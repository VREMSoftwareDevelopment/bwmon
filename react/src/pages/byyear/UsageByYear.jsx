/*
 *      Copyright (C) 2010 - 2025 VREM Software Development <VREMSoftwareDevelopment@gmail.com>
 *
 *      Licensed under the Apache License, Version 2.0 (the "License");
 *      you may not use this file except in compliance with the License.
 *      You may obtain a copy of the License at
 *
 *           http: //www.apache.org/licenses/LICENSE-2.0
 *
 *      Unless required by applicable law or agreed to in writing, software
 *      distributed under the License is distributed on an "AS IS" BASIS,
 *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *      See the License for the specific language governing permissions and
 *      limitations under the License.
 *
 * Bandwidth Monitor
 */

import React from 'react';
import { Paper, Table, TableRow, TableHead, TableContainer } from '@mui/material';
import Body from '../../components/table/Body';
import CellInfo from '../../components/table/CellInfo';
import Header from '../../components/table/Header';
import Pagination from '../../components/table/Pagination';
import { usageInGBytes } from '../../utils/ConversionUtils';
import { comparator, isAscending, sort } from '../../utils/SortUtils';
import useUsageByYear from '../../hooks/byyear/UseUsageByYear';
import { useSortDesc } from '../../hooks/common/UseSort';
import usePagination from '../../hooks/common/UsePagination';
import Loading from '../../components/loading/Loading';

const cellInfos = [
    new CellInfo('id', true, 'left', 'Year', false),
    new CellInfo('download', false, 'right', 'Down', true, usageInGBytes),
    new CellInfo('upload', false, 'right', 'Up', true, usageInGBytes),
    new CellInfo('total', true, 'right', 'Total', true, usageInGBytes),
    new CellInfo('average', false, 'right', 'Average', true, usageInGBytes),
    new CellInfo('days', false, 'left', 'Days', true),
];

const rowsPerPageMin = 20;

const UsageByYear = () => {
    const { data, loading } = useUsageByYear();
    const { page, setPage, rowsPerPage, setRowsPerPage } = usePagination(rowsPerPageMin);
    const { ascending, setAscending, orderBy, setOrderBy } = useSortDesc(cellInfos[0].id);

    const handlePageChange = (event, newPage) => setPage(newPage);

    const handleRowsPerPageChange = (event) => {
        setRowsPerPage(Number(event.target.value));
        setPage(0);
    };

    const handleRequestSort = (event, property) => {
        setAscending(isAscending(orderBy, property, ascending));
        setOrderBy(property);
    };

    const sortedData = () => sort(data, comparator(ascending, orderBy)).slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage);

    const displayData = () =>
        data ? (
            <Table stickyHeader size="small">
                <TableHead>
                    <TableRow>
                        <Pagination
                            data-testid="year-pagination-id"
                            id="year-pagination-id"
                            colSpan={cellInfos.length}
                            count={data.length}
                            minimum={rowsPerPageMin}
                            rowsPerPage={rowsPerPage}
                            page={page}
                            onPageChange={handlePageChange}
                            onRowsPerPageChange={handleRowsPerPageChange}
                        />
                    </TableRow>
                </TableHead>
                <Header
                    prefix="year"
                    cellInfos={cellInfos}
                    onRequestSort={handleRequestSort}
                    ascending={ascending}
                    orderBy={orderBy}
                />
                <Body prefix="year" cellInfos={cellInfos} values={sortedData()} />
            </Table>
        ) : null;

    return (
        <Paper>
            <Loading isLoading={loading} />
            <TableContainer>{displayData()}</TableContainer>
        </Paper>
    );
};

export default UsageByYear;
