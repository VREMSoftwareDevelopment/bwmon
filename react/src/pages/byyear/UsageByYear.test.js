/*
 *      Copyright (C) 2010 - 2020 VREM Software Development <VREMSoftwareDevelopment@gmail.com>
 *
 *      Licensed under the Apache License, Version 2.0 (the "License");
 *      you may not use this file except in compliance with the License.
 *      You may obtain a copy of the License at
 *
 *           http: //www.apache.org/licenses/LICENSE-2.0
 *
 *      Unless required by applicable law or agreed to in writing, software
 *      distributed under the License is distributed on an "AS IS" BASIS,
 *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *      See the License for the specific language governing permissions and
 *      limitations under the License.
 *
 * Bandwidth Monitor
 */

import React from 'react';
import UsageByYear from './UsageByYear';
import { create } from 'react-test-renderer';
// import { act, create } from 'react-test-renderer';
// import wait from '../../__test__/utils/Wait';

jest.mock('../../services/Usage');

describe('UsageByYear', () => {
    test('renders correctly', () => {
        const tree = create(<UsageByYear />);
        expect(tree.toJSON()).toMatchSnapshot();
    });

    // current issue in Material UI Table Pagination label id
    // test('renders correctly after data load', async () => {
    //     let tree;
    //     act(() => {
    //         tree = create(<UsageByYear />);
    //     });
    //     await wait(1);
    //     expect(tree.toJSON()).toMatchSnapshot();
    // });
});
