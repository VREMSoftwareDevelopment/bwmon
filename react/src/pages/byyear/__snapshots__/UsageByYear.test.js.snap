// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsageByYear renders correctly 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    className="makeStyles-root-1"
  >
    <div
      className="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-indeterminate"
      role="progressbar"
    >
      <div
        className="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Indeterminate"
        style={Object {}}
      />
      <div
        className="MuiLinearProgress-bar MuiLinearProgress-bar2Indeterminate MuiLinearProgress-barColorPrimary"
        style={Object {}}
      />
    </div>
    <div
      className="makeStyles-root-2"
    >
      <div
        className="MuiPaper-root MuiAlert-root MuiAlert-standardInfo MuiPaper-elevation0"
        role="alert"
      >
        <div
          className="MuiAlert-icon"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20, 12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10, 10 0 0,0 12,2M11,17H13V11H11V17Z"
            />
          </svg>
        </div>
        <div
          className="MuiAlert-message"
        >
          <div
            className="MuiTypography-root MuiAlertTitle-root MuiTypography-body1 MuiTypography-gutterBottom"
          >
            Loading...
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="MuiTableContainer-root"
  />
</div>
`;
